runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -4788,341 +4800,313 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+   cursor = generateRXInstruction(cg, TR::InstOpCode::LG, node, scratchReg, generateS390MemoryReference(dimsPtrReg, 0, cg), cursor);
+   iComment("Start first dim alloc");
+   // Fist dim class and length:
+   bool useCompRefs = TR::Compiler->om.compressObjectReferences();
+   cursor = generateRXInstruction(cg, useCompRefs ? TR::InstOpCode::ST : TR::InstOpCode::STG, node, classReg,
+      generateS390MemoryReference(resultReg, (int32_t)TR::Compiler->om.offsetOfObjectVftField(), cg), cursor);
+   cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, scratchReg, generateS390MemoryReference(resultReg,
+      (int32_t)TR::Compiler->om.offsetOfContiguousArraySizeField(), cg), cursor);
+
+   /********************************************* Add leaf arrays (Second dimension) *********************************************/
+   // If number of leaf arrays is zero, jump to the end.
+   TR::LabelSymbol *controlFlowEndLabel = generateLabelSymbol(cg);
+   controlFlowEndLabel->setEndInternalControlFlow();
+   cursor = generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CL, node, scratchReg, 0, TR::InstOpCode::COND_BE, controlFlowEndLabel, false /* needsCC */, cursor);
+
+   cursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, dim1SizeReg, generateS390MemoryReference(dim1SizeReg, resultReg, 0, cg), cursor);
+   iComment("dim1SizeReg points to the fist leaf array.");
+   // Load component class to class register. In case of comp refs, class is in high order!
+   cursor = generateRXInstruction(cg, TR::InstOpCode::LG, node, classReg, generateS390MemoryReference(classReg, offsetof(J9ArrayClass, componentType)+(useCompRefs ? 4 : 0), cg), cursor);
+   int32_t shiftAmount = TR::Compiler->om.compressedReferenceShift();
+   if (useCompRefs)
       {
-      // Clear padding in Discontiguous array header layout. Since size field is 4 bytes wide, adding 4 to size field gets us to the padding.
-      cursor = generateSILInstruction(cg, TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField() + 4, cg), 0);
-      iComment("Clear padding.");
-      }
-
-#if defined(J9VM_GC_SPARSE_HEAP_ALLOCATION)
-      if (isOffHeapAllocationEnabled)
+      // Use the class register for both class and length fields.
+      cursor = generateRXInstruction(cg, TR::InstOpCode::L, node, classReg, generateS390MemoryReference(dimsPtrReg, 0, cg), cursor);
+      if (shiftAmount > 0)
          {
-         // Since 1st dimension is 0 length, we can use firstDimLenReg to clear
-         // dataAddr field. secondDimLenReg is expected to be NULL at this point.
-         cursor = generateRXInstruction(cg, TR::InstOpCode::STG, node, firstDimLenReg, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousDataAddrField(), cg));
-         iComment("Clear 1st dim dataAddr field.");
+         // Keep a compressed version of dim2 size in higher 32 bits.
```

Comment:
This seems like an odd comment. It suggest compressed version of dim2 size not the compressed refs. Please update the comment to make it clear. Also having some layout information here would enhance the comment.